File: optionals/med-4.0.0/tests/f/test15.f90
!*  This file is part of MED.
!*
!*  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
!*  MED is free software: you can redistribute it and/or modify
!*  it under the terms of the GNU Lesser General Public License as published by
!*  the Free Software Foundation, either version 3 of the License, or
!*  (at your option) any later version.
!*
!*  MED is distributed in the hope that it will be useful,
!*  but WITHOUT ANY WARRANTY; without even the implied warranty of
!*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!*  GNU Lesser General Public License for more details.
!*
!*  You should have received a copy of the GNU Lesser General Public License
!*  along with MED.  If not, see <http://www.gnu.org/licenses/>.
!*

! *******************************************************************************
! * - Nom du fichier : test15.f90
! *
! * - Description : lecture des noeuds d'un maillage MED.
! *                 a l'aide des routines de niveau 2
! *                 - equivalent a test5.f90
! *
! ******************************************************************************

program test15
  
  implicit none
  include 'med.hf90'
!  
!
  integer*8  fid
  integer  ret,cret
  !  ** la dimension du maillage                        **
  integer mdim,sdim
  !  ** nom du maillage de longueur maxi MED_TAILLE_NOM **
  character*64 maa
  character*200 desc
  !  ** le nombre de noeuds                             **
  integer :: nnoe = 0
  !  ** table des coordonnees                           **
  real*8, allocatable, dimension(:) :: coo
  !  ** tables des noms et des unites des coordonnees 
  !     profil : (dimension)                            **
  character*16 nomcoo(2)
  character*16 unicoo(2)
  character*16 dtunit
  !  ** tables des noms, numeros, numeros de familles des noeuds
  !     autant d'elements que de noeuds - les noms ont pout longueur
  !     MED_SNAME_SIZE **
  character*16, allocatable, dimension(:) ::  nomnoe
  integer,      allocatable, dimension(:) ::  numnoe,nufano
  integer rep
  integer inonoe,inunoe,inufa
  character*16 str
  integer i
  character*255 argc
  integer type,nstep,stype
  integer chgt,tsf

  !  ** Ouverture du fichier **
  call mfiope(fid,"test14.med",MED_ACC_RDONLY, cret)
  print *,cret
   

  !  ** Lecture des infos concernant le premier maillage **
  if (cret.eq.0) then
     call mmhmii(fid,1,maa,sdim,mdim,type,desc,dtunit,stype,nstep,rep,nomcoo,unicoo,cret)
     print *,"Maillage de nom : ",maa," et de dimension : ",mdim
  endif
  print *,cret
  
  ! ** Lecture du nombre de noeud **
  if (cret.eq.0) then
     call mmhnme(fid,maa,MED_NO_DT,MED_NO_IT,MED_NODE,MED_NONE,MED_COORDINATE,MED_NO_CMODE,chgt,tsf,nnoe,cret)
     print *,"Nombre de noeuds : ",nnoe
  endif
  print *,cret

  ! ** Allocations memoires **
  ! ** table des coordonnees 
  ! ** profil : (dimension * nombre de noeuds ) **
  allocate (coo(nnoe*sdim),STAT=ret)
  ! ** table des des numeros, des numeros de familles des noeuds
  !   profil : (nombre de noeuds) **
  allocate (numnoe(nnoe),nufano(nnoe),STAT=ret)
  ! ** table des noms des noeuds 
  !   profil : (nnoe*MED_TAILLE_PNOM+1) **
  allocate (nomnoe(nnoe),STAT=ret)

  ! ** Lecture des noeuds : 
  !     - Coordonnees
  !     - Noms (optionnel dans un fichier MED) 
  !     - Numeros (optionnel dans un fichier MED) 
  !     - Numeros de familles	**
  if (cret.eq.0) then
     call mmhnor(fid,maa,MED_NO_DT,MED_NO_IT,MED_FULL_INTERLACE,coo,inonoe,nomnoe,inunoe,numnoe,inufa,nufano,cret)    
  endif
  
  ! ** Affichage des resulats **
  if (cret.eq.0) then
      print *,"Type de repere : ",rep
      print *,"Nom des coordonnees : ",nomcoo
    
      print *,"Unites des coordonnees : ",unicoo
     
      print *,"Coordonnees des noeuds : ",coo
     
      if (inonoe .eq. MED_TRUE) then
         print *,"Noms des noeuds : |",nomnoe,"|"
      endif

      if (inunoe .eq. MED_TRUE) then
         print *,"Numeros des noeuds : ",numnoe
      endif

      if (inufa .eq. MED_TRUE) then
         print *,"Numeros des familles des noeuds : ",nufano
      else
         print *,"Numeros des familles des noeuds : 0"
      endif

   endif

  ! ** Liberation memoire **
   deallocate(coo,nomnoe,numnoe,nufano)
 
  ! ** Fermeture du fichier **
   call mficlo(fid,cret)
   print *,cret

  ! **Code retour
   call efexit(cret)
   
 end program test15

File: optionals/med-4.0.0/tests/c/test15.c
/*  This file is part of MED.
 *
 *  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
 *  MED is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  MED is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with MED.  If not, see <http://www.gnu.org/licenses/>.
 */

/******************************************************************************
 * - Nom du fichier : test15.c
 *
 * - Description : lecture des noeuds d'un maillage MED
 *                 a l'aide des routines de niveau 2
 *                 - equivalent a test5.c
 *
 *****************************************************************************/

#include <med.h>
#define MESGERR 1
#include "med_utils.h"
#include <string.h>

#ifdef DEF_LECT_ECR
#define MODE_ACCES MED_ACC_RDWR
#elif DEF_LECT_AJOUT
#define MODE_ACCES MED_ACC_RDEXT
#else
#define MODE_ACCES MED_ACC_CREAT
#endif

int main (int argc, char **argv)


{
  med_err ret = 0;
  med_idt fid = 0;
  med_int mdim=0,sdim=0;
  /* nom du maillage de longueur maxi MED_NAME_SIZE */
  char maa[MED_NAME_SIZE+1];
  /* le nombre de noeuds */
  med_int nnoe = 0;
  /* table des coordonnees */
  med_float *coo;
  /* tables des noms et des unites des coordonnees (dimension*MED_SNAME_SIZE+1) */
  char nomcoo[3*MED_SNAME_SIZE+1]="";
  char unicoo[3*MED_SNAME_SIZE+1]="";
  /* tables des noms, numeros, numeros de familles des noeuds
     autant d'elements que de noeuds - les noms ont pour longueur  MED_SNAME_SIZE */
  char       *nomnoe;
  med_int    *numnoe;
  med_int    *nufano;
  med_bool   inonoe,inunoe,inufam;
  med_bool   chgt=MED_FALSE, trsf=MED_FALSE;
  char str   [MED_SNAME_SIZE+1]  ="";
  char desc  [MED_COMMENT_SIZE+1]="";
  char dtunit[MED_SNAME_SIZE+1]  ="";
  med_mesh_type    type;
  med_sorting_type sort;
  med_axis_type    rep;
  med_int          nstep=0;
  med_int i;

  /* Ouverture du fichier en mode lecture seule */
  if ((fid = MEDfileOpen(argv[1],MED_ACC_RDONLY)) < 0) {
    MESSAGE("Erreur a l'ouverture du fichier.");
    return -1;
  }
  if ((sdim=MEDmeshnAxis(fid, 1)) <0) {
    MESSAGE("Erreur a la lecture de la dimension de l'espace du maillage :");
    SSCRUTE(maa);
    return -1;
  }

  /* Lecture des infos concernant le premier maillage */
  if ( MEDmeshInfo( fid, 1,  maa, &sdim, &mdim, &type, desc, dtunit, &sort,
		    &nstep,  &rep, nomcoo,unicoo) < 0 ) {
    MESSAGE("Erreur a la lecture des informations sur le maillage : ");SSCRUTE(maa);
    return -1;
  } else {
    printf("Maillage de nom : |%s| , de dimension : "IFORMAT" , et de type %d\n",maa,mdim,type);
    printf("\t -Dimension de l'espace : "IFORMAT"\n",sdim);
    printf("\t -Description du maillage : |%s|\n",desc);
    printf("\t -Noms des axes : |%s|\n",nomcoo);
    printf("\t -Unités des axes : |%s|\n",unicoo);
    printf("\t -Type de repère : %d\n",rep);
    printf("\t -Nombre d'étapes de calcul : "IFORMAT"\n",nstep);
    printf("\t -Unité des dates : |%s|\n",dtunit);
  }

  /* Combien de triangles et de segments */
  if ((nnoe = MEDmeshnEntity(fid, maa, MED_NO_DT, MED_NO_IT,
			      MED_NODE, MED_NODE,MED_COORDINATE, MED_NO_CMODE,
			     &chgt, &trsf)) < 0)  {
    MESSAGE("Erreur a la lecture du nombre de noeuds");
    return -1;
  }
  printf("Nombre de noeuds : "IFORMAT" \n",nnoe);

  /* Allocations memoires */
  /* table des coordonnees
     profil : (dimension * nombre de noeuds ) */
  if (nnoe > 0) {
    coo = (med_float*) malloc(sizeof(med_float)*nnoe*mdim);
    /* table des des numeros, des numeros de familles des noeuds
       profil : (nombre de noeuds) */
    numnoe = (med_int*) malloc(sizeof(med_int)*nnoe);
    nufano = (med_int*) calloc(nnoe,sizeof(med_int));
    /* table des noms des noeuds
       profil : (nnoe*MED_SNAME_SIZE+1) */
    nomnoe = (char*) malloc(MED_SNAME_SIZE*nnoe+1);

    /* Lecture des noeuds :
       - Coordonnees
       - Noms (optionnel dans un fichier MED)
       - Numeros (optionnel dans un fichier MED)
       - Numeros de familles	*/
    if (MEDmeshNodeRd(fid,maa,MED_NO_DT,MED_NO_IT,MED_FULL_INTERLACE,
		      coo,&inonoe,nomnoe,&inunoe,numnoe,&inufam,nufano) < 0) {
      MESSAGE("Erreur a la lecture des noeuds du maillage");
      ret = -1;
    }

    /* Affichage */
    if (ret == 0) {
      printf("Type de repere : %d \n",rep);
      printf("Nom des coordonnees : \n");
      for (i=0;i<mdim;i++) {
	strncpy(str,nomcoo+i*MED_SNAME_SIZE,MED_SNAME_SIZE);
	str[MED_SNAME_SIZE] = '\0';
	printf("|%s| ",str);
      }
      printf("\nUnites des coordonnees : \n");
      for (i=0;i<mdim;i++) {
	strncpy(str,unicoo+i*MED_SNAME_SIZE,MED_SNAME_SIZE);
	str[MED_SNAME_SIZE] = '\0';
	printf("|%s| ",str);
      }
      printf("\nCoordonnees des noeuds : \n");
      for (i=0;i<nnoe*mdim;i++)
	printf("%f ",*(coo+i));
      if (inonoe) {
	printf("\nNoms des noeuds : \n");
	for (i=0;i<nnoe;i++) {
	  strncpy(str,nomnoe+i*MED_SNAME_SIZE,MED_SNAME_SIZE);
	  str[MED_SNAME_SIZE] = '\0';
	  printf(" |%s| ",str);
	}
      }
      if (inunoe) {
	printf("\nNumeros des noeuds : \n");
	for (i=0;i<nnoe;i++)
	  printf(""IFORMAT" ",*(numnoe+i));
      }

      printf("\nPrésence de numeros des familles des noeuds : %d\n",inufam);
      printf("\nNumeros des familles des noeuds : \n");
      for (i=0;i<nnoe;i++)
	printf(IFORMAT" ",*(nufano+i));
      printf("\n");
    }

    /* Liberation memoire */
    free(coo);
    free(nomnoe);
    free(numnoe);
    free(nufano);
  }

  /* Fermeture du fichier */
  if (MEDfileClose(fid) < 0) {
    MESSAGE("Erreur a la fermeture du fichier");
    return -1;
  }

  return ret;
}




